### ISSUE TITLE
Importing CKEditor in an isomorphic app

### ISSUE THREAD

**Issue by arnaudriegert** [ckeditor/ckeditor5-react#31]
I'm building an isomorphic app in which I want to integrate CKEditor.

CKEditor needs `window` to work and therefore it cannot be initialized on the Node server.

I ended up defining this component that renders a CKEditor if `window` is defined and a `textarea` otherwise.

```jsx
import React from 'react';

class RichTextEditor extends React.Component {
  componentDidMount () {
    // TODO Find out if we have a textarea or a CKEditor
    // If it is a textarea, try rendering once more (we should be in the browser by now)
  }
  
  render () {
    // CKEditor depends on window and therefore it cannot be initialized server-side.
    // We show a basic textarea as a fallback.
    if (typeof window === 'undefined') {
      return (
        <textarea {...this.props} />
      );
    } else {
      // Cannot import at the beginning of the file, will crash with ReferenceError: window is not defined
      const CKEditor = require('@ckeditor/ckeditor5-react');
      const ClassicEditor = require('@ckeditor/ckeditor5-build-classic');
      return (
        <CKEditor editor={ ClassicEditor } {...this.props} />
      );
    }
  }
}

export default RichTextEditor;
```

I cannot `import CKEditor` (or `require` it) at the beginning of the file or the server will crash, so I `required` it from within the `render` function once I know that `window` exists. 
Does that seem like a valid approach? Is it by design that I cannot `import CKEditor` at all if `window` isn't defined?

**Comment by Reinmar** [comment]
cc @pomek @ma2ciek

**Comment by ma2ciek** [comment]
Hi, @arnaudriegert.

In contrast to the `react` library, CKEditor5 builds come with a builtin UI stuff - DOM listeners etc. There was an idea in the past to create a [headless editor](https://github.com/ckeditor/ckeditor5/issues/438), but due to lack of interest, we didn't go further on this topic. And I'm not sure if it would be a good solution for you, because you would still need two different builds for two different environments.

If we fix all browser-related requirements, then it would be possible to import the editor build only once. But the `<CKEditor>` component with ClassicEditor actually modifies the DOM, attaches DOM listeners and so on. I guess it would break in the virtual DOM, so it would only allow having one import at the top and still an `if` statement in the code.

Eventually, the `<CKEditor>` component could render a nice and styled template if the component detects the node env, but I'm not sure if this can be easily achievable.

**Comment by Reinmar** [comment]
> There was an idea in the past to create a headless editor, but due to lack of interest, we didn't go further on this topic.

To clarify – we made some architecture changes to make that easily achievable. There's a chance that running a headless editor in Node.js would work even now, but we didn't check because of a low interest in this topic.

> If we fix all browser-related requirements, then it would be possible to import the editor build only once.

That sounds doable. I think we could make CKEditor 5 importable, so there are no static-level incompatibilities. But I totally agree that starting an entire CKEditor 5 build in Node.js doesn't make sense. It won't serve any purpose because there's no chance that it will speed up the load times of the final web page.

So, I'd recommend adding a check which conditionally stops enabling the editor. Alternatively, you could try to conditionally use a headless editor in that component but that's far more work because you'd have to create that headless editor class too.

On our side, we'll have to deal with static dependency on browser's globals. Perhaps you'd like to check, @arnaudriegert, in which places we use `window` and e.g. `document` in modules' main scopes (so the code is executed even though you didn't try to start the editor)? That may speed things up.

Alternatively, you could switch from `import X from` to dynamic imports (`import()` or `require()`). The goal would be to not import CKEditor 5 at all if not in a browser.

**Comment by Reinmar** [comment]
> If we can import CKEditor in the main scope even if window isn't defined, we take the risk of having a crash at runtime, since you're going to need to access the DOM when you init your editor anyway, right?

Yes.

> If you prevent your app from running at all with a main scope import you'll force developers to check if window exists before they try to init.

That I don't understand. If we prevent running CKE at all with a main scope import, then nothing works because the app containing CKE crashes before you're able to check if `window` exists. That's why I'd like if you could import CKE and check later if you can init the editor.

> Unless there is another advantage I'm not aware of in using static imports, I wouldn't be shocked if it remained as it is now.

Static imports allow more optimisations during the building process. But if your app wants to e.g. delay loading the editor at all (aka dynamic imports), then AFAIK it cannot use static imports anyway.

In other words, if you can use dynamic imports in your app, then the problem you reported is of little importance because you may load all CKE5 conditionally. However, I'd like if loading it statically worked anyway, just for the sake of purity :)

**Comment by arnaudriegert** [comment]
When I said 

> you'll force developers to check if window exists before they try to init

I meant 

> you'll force developers to check if window exists before they even import it 


But I see your point :) I'll try to help with that!
